Support for downloading on-device models

teelisyys opened the pull request.

thisismatu commented:
@teelisyys looks good, but few things needs addressing:

1\. When entering an invalid option to the model flag the error message is printed twice along with the help message. One error message should be suffcicient without the help message.

```
> ./bin/speechly download 12345 testi --model asd
Error: "asd" is not a valid option. Available options are: ort, tflite, coreml and all
Usage:
  speechly download [<app_id>] <directory> [flags]

Examples:
speechly download <app_id> /path/to/config
speechly download -a <app_id> .
speechly download -a <app_id> . --model tflite

Flags:
  -a, --app string     Which application's configuration or model to download. Can be given as the first positional argument.
  -h, --help           help for download
      --model string   Specify the machine learning framework model to download. Available options are: ort, tflite, coreml and all. This feature is available on Enterprise plans (https://speechly.com/pricing)

Error executing command: "asd" is not a valid option. Available options are: ort, tflite, coreml and all
```

2\. When entering just some random app_id i get a memory access violation.

```
> ./bin/speechly download 9dbba3e2-0d75-40db-a45a-a351d6cb8ce7 testi --model=all
Downloading to /Users/matu/Sites/speechly/cli/testi
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x30 pc=0x15277d6]

goroutine 1 [running]:
github.com/speechly/cli/cmd.downloadCurrentModel({0x174c268, 0xc0002180f0}, {0xc0000425d0, 0x24}, {0x7ff7bfeffb28, 0x24}, {0x16510b5, 0x3})
	/Users/matu/Sites/speechly/cli/cmd/download.go:177 +0x356
github.com/speechly/cli/cmd.glob..func18(0x1af4660, {0xc000218330, 0x2, 0x3?})
	/Users/matu/Sites/speechly/cli/cmd/download.go:67 +0x287
github.com/spf13/cobra.(*Command).execute(0x1af4660, {0xc000218300, 0x3, 0x3})
	/Users/matu/go/pkg/mod/github.com/spf13/cobra@v1.3.0/command.go:860 +0x663
github.com/spf13/cobra.(*Command).ExecuteC(0x1af5ce0)
	/Users/matu/go/pkg/mod/github.com/spf13/cobra@v1.3.0/command.go:974 +0x3bd
github.com/spf13/cobra.(*Command).Execute(...)
	/Users/matu/go/pkg/mod/github.com/spf13/cobra@v1.3.0/command.go:902
github.com/spf13/cobra.(*Command).ExecuteContext(...)
	/Users/matu/go/pkg/mod/github.com/spf13/cobra@v1.3.0/command.go:895
github.com/speechly/cli/cmd.Execute()
	/Users/matu/Sites/speechly/cli/cmd/root.go:41 +0x65
main.main()
	/Users/matu/Sites/speechly/cli/main.go:10 +0x19
```

thisismatu reviewed (changes requested):
@teelisyys looks good, but few things needs addressing:

1\. When entering an invalid option to the `model` flag the error message is printed twice along with the help message. One error message should be sufficient without the help message.

```
> ./bin/speechly download 12345 testi --model asd
Error: "asd" is not a valid option. Available options are: ort, tflite, coreml and all
Usage:
  speechly download [<app_id>] <directory> [flags]

Examples:
speechly download <app_id> /path/to/config
speechly download -a <app_id> .
speechly download -a <app_id> . --model tflite

Flags:
  -a, --app string     Which application's configuration or model to download. Can be given as the first positional argument.
  -h, --help           help for download
      --model string   Specify the machine learning framework model to download. Available options are: ort, tflite, coreml and all. This feature is available on Enterprise plans (https://speechly.com/pricing)

Error executing command: "asd" is not a valid option. Available options are: ort, tflite, coreml and all
```

2\. When entering just some random `app_id` i get a memory access violation.

```
> ./bin/speechly download 9dbba3e2-0d75-40db-a45a-a351d6cb8ce7 testi --model=all
Downloading to /Users/matu/Sites/speechly/cli/testi
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x30 pc=0x15277d6]

goroutine 1 [running]:
github.com/speechly/cli/cmd.downloadCurrentModel({0x174c268, 0xc0002180f0}, {0xc0000425d0, 0x24}, {0x7ff7bfeffb28, 0x24}, {0x16510b5, 0x3})
	/Users/matu/Sites/speechly/cli/cmd/download.go:177 +0x356
github.com/speechly/cli/cmd.glob..func18(0x1af4660, {0xc000218330, 0x2, 0x3?})
	/Users/matu/Sites/speechly/cli/cmd/download.go:67 +0x287
github.com/spf13/cobra.(*Command).execute(0x1af4660, {0xc000218300, 0x3, 0x3})
	/Users/matu/go/pkg/mod/github.com/spf13/cobra@v1.3.0/command.go:860 +0x663
github.com/spf13/cobra.(*Command).ExecuteC(0x1af5ce0)
	/Users/matu/go/pkg/mod/github.com/spf13/cobra@v1.3.0/command.go:974 +0x3bd
github.com/spf13/cobra.(*Command).Execute(...)
	/Users/matu/go/pkg/mod/github.com/spf13/cobra@v1.3.0/command.go:902
github.com/spf13/cobra.(*Command).ExecuteContext(...)
	/Users/matu/go/pkg/mod/github.com/spf13/cobra@v1.3.0/command.go:895
github.com/speechly/cli/cmd.Execute()
	/Users/matu/Sites/speechly/cli/cmd/root.go:41 +0x65
main.main()
	/Users/matu/Sites/speechly/cli/main.go:10 +0x19
```

thisismatu commented on cmd/download.go:
```diff
@@ -203,5 +203,5 @@ func downloadCurrentModel(ctx context.Context, absPath string, appId string, mod
 	RootCmd.AddCommand(downloadCmd)
 	downloadCmd.Flags().StringP("app", "a", "", "Which application's configuration or model to download. Can be given as the first positional argument.")
-	downloadCmd.Flags().String("model", "", "Download the model used by the application. Available machine learning frameworks are ort, tflite and coreml, if you specify all, all frameworks available for you will be downloaded. This feature is available on Enterprise plans (https://speechly.com/pricing).")
+	downloadCmd.Flags().String("model", "", "Specify the machine learning framework model to download. Available options are: ort, tflite, coreml and all. This feature is available on Enterprise plans (https://speechly.com/pricing)")
```
@aukkonen @teelisyys the original help text felt clumsy after the change to how the model flag is used, so i rewrote it. Do you think this makes sense? Is "Specify the _machine learning framework model_ to download" the correct thing to say?

teelisyys replied:
I'd use "_...machine learning framework of the model..._" for extra clarity

teelisyys commented:
@thisismatu 

1. This `error+usage+error` seems to be consistent across the app for the param validation. I've now removed the second error but kept the usage, as not all error messages currently include the information on what is expected.
2. This is now fixed.
